src/main/java/org/example/jvspringbootfirstbook/security/token/JwtAuthenticationFilter.java

```diff
@@ -0,0 +1,54 @@
+package org.example.jvspringbootfirstbook.security.token;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private final JwtUtil jwtUtil;
+    private final UserDetailsService userDetailsService;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request,
+                                    HttpServletResponse response,
+                                    FilterChain filterChain) throws ServletException, IOException {
+
+        String tokenFromRequest = getTokenFromRequest(request);
+
+        if (tokenFromRequest != null && jwtUtil.isValid(tokenFromRequest)) {
+
+            String username = jwtUtil.getUserName(tokenFromRequest);
+            UserDetails userDetails = userDetailsService.loadUserByUsername(username);
```

Comment:
Think if it's possible that userDetails will be null? If so, add a handler for that case